### PR TITLE
Implement new style softmax

### PR DESCRIPTION
**Pull request by takagi**
Rel #3147.

This PR implements the new style softmax layer along with double backward support.

### CODE REVIEW

**Comment by takagi** [comment]
Thanks to @beam2d , now I can remove WIP.

File: chainer/functions/activation/softmax.py [hunk 1]
```diff
@@ -2,7 +2,9 @@
 
 import chainer
 from chainer import cuda
-from chainer import function
+from chainer import function_node
+from chainer.functions.array import broadcast
```
Comment (beam2d) [inline chainer/functions/activation/softmax.py]: Use `chainer.functions` namespace instead (same for the next line).

File: chainer/functions/activation/softmax.py [hunk 2]
```diff
@@ -50,12 +60,23 @@ def forward(self, x):
             y /= y.sum(axis=self.axis, keepdims=True)
 
         self._x_shape = x[0].shape
```
Comment (beam2d) [inline chainer/functions/activation/softmax.py]: It looks you can remove it.

File: chainer/functions/activation/softmax.py [hunk 3]
```diff
@@ -64,26 +85,35 @@ def backward(self, x, gy):
...
-        left_shape = numpy.prod(shape[slice(0, self.axis)], dtype=numpy.int)
-        center_shape = shape[self.axis]
-        right_shape = numpy.prod(
-            shape[slice(self.axis + 1, len(shape))], dtype=numpy.int)
-        return left_shape, center_shape, right_shape, 1
+    def backward(self, indexes, grad_outputs):
+        y, gy = self.get_retained_inputs()
+        ggx, = grad_outputs
+        gs = sum_.sum(-ggx * y, axis=self.axis, keepdims=True)
+        ga = ggx + broadcast.broadcast_to(gs, gy.shape)
```
Comment (beam2d) [inline chainer/functions/activation/softmax.py]: It's better to subtract `sum(ggx*y)` instead of adding `sum(-ggx*y)`. The number of function applications will be reduced by one.

File: chainer/functions/activation/softmax.py [hunk 4]
```diff
@@ -64,26 +85,35 @@ def backward(self, x, gy):
...
+        ret = []
+        if 0 in indexes:
+            s = broadcast.broadcast_to(
+                sum_.sum(y * gy, axis=self.axis, keepdims=True), gy.shape)
+            gy2 = -ggx * s + ga * gy
```
Comment (beam2d) [inline chainer/functions/activation/softmax.py]: Better to write `ga * gy - ggx * s`.

File: chainer/functions/activation/softmax.py (hunk 1 above)
Comment (takagi) [reply chainer/functions/activation/softmax.py]: Replaced!

File: chainer/functions/activation/softmax.py (hunk 2 above)
Comment (takagi) [reply chainer/functions/activation/softmax.py]: Removed!

File: chainer/functions/activation/softmax.py (hunk 3 above)
Comment (takagi) [reply chainer/functions/activation/softmax.py]: Move the subtract out!

File: chainer/functions/activation/softmax.py (hunk 4 above)
Comment (takagi) [reply chainer/functions/activation/softmax.py]: Fixed the order!

File: chainer/functions/activation/softmax.py [hunk 5]
```diff
@@ -64,26 +83,35 @@ def backward(self, x, gy):
...
-        left_shape = numpy.prod(shape[slice(0, self.axis)], dtype=numpy.int)
-        center_shape = shape[self.axis]
-        right_shape = numpy.prod(
-            shape[slice(self.axis + 1, len(shape))], dtype=numpy.int)
-        return left_shape, center_shape, right_shape, 1
+    def backward(self, indexes, grad_outputs):
+        y, gy = self.get_retained_inputs()
+        ggx, = grad_outputs
+        gs = -chainer.functions.sum(ggx * y, axis=self.axis, keepdims=True)
+        ga = ggx + chainer.functions.broadcast_to(gs, gy.shape)
```
Comment (beam2d) [inline chainer/functions/activation/softmax.py]: Sorry, I was ambiguous. I though it's better to write it as follows.
```
gs = chainer.functions.sum(...)
ga = ggx - chainer.functions.broadcast_to(gs, ...)
```
so that negation+add (two functions) is replaced by subtract (one function).

File: tests/chainer_tests/functions_tests/activation_tests/test_softmax.py [hunk 6]
```diff
@@ -101,6 +104,33 @@ def test_backward_gpu_non_contiguous(self):
...
 
+    def check_double_backward(self, x_data, gy_data, ggx_data,
+                              use_cudnn='always'):
+        def f(x):
+            return functions.softmax(x, axis=self.axis)
+
+        with chainer.using_config('use_cudnn', use_cudnn):
+            gradient_check.check_double_backward(
+                f, (x_data,), (gy_data,), (ggx_data),
+                **self.check_backward_options)
```
Comment (beam2d) [inline tests/chainer_tests/functions_tests/activation_tests/test_softmax.py]: Sorry, one another comment: it seems this test is numerically unstable. I recommend to use a slightly looser threshold for double backprop test here.

File: chainer/functions/activation/softmax.py (hunk 5 above)
Comment (takagi) [reply chainer/functions/activation/softmax.py]: Yes, fixed!

File: tests/chainer_tests/functions_tests/activation_tests/test_softmax.py (hunk 6 above)
Comment (takagi) [reply tests/chainer_tests/functions_tests/activation_tests/test_softmax.py]: I fixed to give looser threshold by an order for double backprop test in case that dtype is float16. Is it enough or better to fix for other dtypes too?

**Comment by beam2d** [comment]
LGTM!
